packages/block-library/src/categories/edit.js

```diff
@@ -30,23 +29,14 @@ export default function CategoriesEdit( {
 	setAttributes,
 } ) {
 	const selectId = useInstanceId( CategoriesEdit, 'blocks-category-select' );
-	const { categories, isRequesting } = useSelect(
-		( select ) => {
-			const { getEntityRecords, isResolving } = select( coreStore );
-			const query = { per_page: -1, hide_empty: true, context: 'view' };
-			if ( showOnlyTopLevel ) {
-				query.parent = 0;
-			}
-			return {
-				categories: getEntityRecords( 'taxonomy', 'category', query ),
-				isRequesting: isResolving( 'getEntityRecords', [
-					'taxonomy',
-					'category',
-					query,
-				] ),
-			};
-		},
-		[ showOnlyTopLevel ]
+	const query = { per_page: -1, hide_empty: true, context: 'view' };
+	if ( showOnlyTopLevel ) {
+		query.parent = 0;
+	}
+	const { categories, isResolving } = useEntityRecords(
+		'taxonomy',
+		'category',
+		query
```

Comment:
Just double checking here. On every re-render, we will have a different query object here, but it shouldn't matter because `@wordpress/core-date` picks data based on the standardised key computed from the query, right? I guess what I mean here is that before `useSelect` would get executed only when `showOnlyTopLevel` changes or when the store triggers an update. However, with the new implementation, it has to run `useEntityRecords` every time we pass a new `query` object.

I believe the same concern applies to every usage of `useEntityRecords` where `query` is dynamically computed which is all cases in this PR. We should check if there are any performance implications here.

Reply:
AFAIR deps are tested for shallow equality, so it doesn't matter as long as there are no nested objects such as `{ _fields: [ 'id' ] }`. I such a case, it would re-run the select callback, which then will return the exact same data.

This may or may not have performance implications, depending on the specifics. I'd like to make sure this just never re-runs. I have two ideas:

1. Add useMemo() in all such places
2. Don't include queryArgs in deps and instead, rely on `JSON.stringify( queryArgs )` – this removes the need to remember about useMemo() and should still be fast.

Reply:
Instead of trying to guess whether the query has changed or not, I added an additional `deps` argument.

Reply:
I would recommend not introducing new APIs with custom dependencies array, as [suggested](https://www.npmjs.com/package/eslint-plugin-react-hooks#advanced-configuration) by the React team (with that in mind, we should probably refactor `useQuerySelect` at some point):

> Generally saying, we recommend most custom Hooks to not use the dependencies argument, and instead provide a higher-level API that is more focused around a specific use case.

Serializing the query into a stabilized key is a possible solution. I believe `getEntityRecords` has already done that internally as @gziolo mentioned. We could do similar things here, maybe implicitly in `useQuerySelect`? I think this is how `react-query` does it too, but they leave this option to the users to provide their own stabilized key for each query.